src/layers/transform/evaluation.cpp

```diff
@@ -70,85 +72,104 @@ void fp_gpu(lbann_comm& comm,
   // GPU objects
   GPUMat sum_d, ones_d;
 #ifdef HYDROGEN_HAVE_CUB
-  sum_d.SetMemoryMode(1);  // Use CUB GPU memory pool if possible
-  ones_d.SetMemoryMode(1); // Use CUB GPU memory pool if possible
+  sum_d.SetMemoryMode(1);  // Use CUB GPU memory pool
+  ones_d.SetMemoryMode(1); // Use CUB GPU memory pool
 #endif // HYDROGEN_HAVE_CUB
+  sum_d.Resize(1, 1);
   auto&& handle = El::GPUManager::cuBLASHandle();
+  auto&& stream = El::GPUManager::Stream();
   CHECK_CUBLAS(cublasSetPointerMode(handle, CUBLAS_POINTER_MODE_DEVICE));
 
   // Compute sum of local input matrix entries
-  if (local_height < 1 || local_width < 1) {
-    El::Zeros(sum_d, 1, 1);
-  } else if (local_height == local_input.LDim() || local_width == 1) {
-    sum_d.Resize(1, 1);
+  if (local_input.IsEmpty()) {
+    El::Zero(sum_d);
+  } else if (local_input.Contiguous()) {
     ones_d.Resize(local_height * local_width, 1);
-    El::Fill(ones_d, DataType(1));
+    El::Fill(ones_d, one);
     cublas::dot(handle,
                 local_height * local_width,
                 local_input.LockedBuffer(), 1,
                 ones_d.LockedBuffer(), 1,
                 sum_d.Buffer());
   } else if (local_height == 1) {
-    sum_d.Resize(1, 1);
     ones_d.Resize(local_width, 1);
-    El::Fill(ones_d, DataType(1));
+    El::Fill(ones_d, one);
     cublas::dot(handle,
                 local_width,
                 local_input.LockedBuffer(), local_input.LDim(),
                 ones_d.LockedBuffer(), 1,
                 sum_d.Buffer());
   } else {
-    sum_d.Resize(local_width + 1, 1);
-    ones_d.Resize(std::max(local_height, local_width), 1);
-    El::Fill(ones_d, DataType(1));
-    for (El::Int col = 0; col < local_width; ++col) {
-      cublas::dot(handle,
-                  local_height,
-                  local_input.LockedBuffer(0, col), 1,
-                  ones_d.LockedBuffer(), 1,
-                  sum_d.Buffer(col+1, 0));
+    GPUMat col_sums_d;
+#ifdef HYDROGEN_HAVE_CUB
+    col_sums_d.SetMemoryMode(1);  // Use CUB GPU memory pool
+#endif // HYDROGEN_HAVE_CUB
+    col_sums_d.Resize(local_width, 1);
+    ones_d.Resize(local_height, 1);
+    El::Fill(ones_d, one);
+    El::Gemv(El::TRANSPOSE, one, local_input, ones_d, zero, col_sums_d);
```

Comment:
Just to clarify, you converted the loop of dot products to a GEMV? This should perform better, but would be good to confirm.

Reply:
That's right. From a latency perspective, it's much better to launch one GEMV kernel than many DOT kernels. I suspect the compute time will be better two since it only needs to read in `ones_d` once.